gtsfm/densify/mvs_math.py

```diff
@@ -0,0 +1,47 @@
+"""MVS math methods for gtsfm
+
+Authors: Ren Liu
+"""
+import math
+import numpy as np
+
+from gtsfm.utils.geometry_comparisons import angle_between_vectors
+
+
+def piecewise_gaussian(
+    a_x: np.ndarray, b_x: np.ndarray, theta_0: float = 5, sigma_1: float = 1, sigma_2: float = 10
+) -> float:
+    """Evaluate the similarity of measurements for the same track in camera a and camera b.
+    1. This piecewise Gaussian function outputs a float score to show the evaluation result of the similarity.
+    2. The total similarity between two views can be calculated by summing up the scores of all common tracks.
+    3. A higher score suggests that the parallax of the track point between camera a and camera b is closer to a small
+    pre-defined angle theta_0 (5 degrees in default), which means views of camera a and camera b have many shared
+    points but with identifiable differences. So the view pair is suitable to be set as the reference view and
+    the source view.
+
+    More details can be found in "View Selection" paragraphs in Yao's paper https://arxiv.org/abs/1804.02505.
+
+    Args:
+        a_x: 3D coordinates of the track point in camera a's frame, with shape (3,).
+        b_x: 3D coordinates of the track point in camera b's frame, with shape (3,).
+        theta_0: Default theta_0 is set to be 5.
+            theta_0 is the threshold angle (in degree) between coordinates in different poses.
+        sigma_1: Default sigma_1 is set to be 1.
+            If the angle between measurements in different views is no larger than the threshold angle, which means
+            view a and b are similar in this track, then the Gaussian variance should be smaller to make the score
+            higher.
+        sigma_2: Default sigma_2 is set to be 10.
+            If the angle between measurements in different views is larger than the threshold angle, which means
+            view a and b are not similar in this track, so less importance should be attached to this track. The
+            Gaussian variance should be larger to make the score lower.
+
+    Returns:
+        A score of the track between two views in the range (0,1]
+    """
+    # 1. calculate the angle between measurement poses of track p in views a and b.
```

Comment:
<img width="595" alt="Screen Shot 2021-05-27 at 1 12 30 AM" src="https://user-images.githubusercontent.com/16724970/119769559-a94a9880-be88-11eb-87c7-7b8aa63c1239.png">

@codyly are you sure this is equivalent? the dot products and quantity inside arccos?

Reply:
@johnwlambert p is the track's coordinates in world frame, ci is the camera center coordinates in world frame, I think it's equivalent if (p-ci) is equal to p's coordinates in camera frame...

Reply:
Thanks -- does Fangjinhua implement it the way as us?